selfdrive/ui/soundd.py

```diff
@@ -0,0 +1,152 @@
+import math
+import time
+import numpy as np
+import wave
+
+from typing import Dict, Optional, Tuple
+
+from cereal import car, messaging
+from openpilot.common.basedir import BASEDIR
+from openpilot.common.realtime import Ratekeeper
+from openpilot.system.hardware import PC
+from openpilot.system.swaglog import cloudlog
+
+SAMPLE_RATE = 48000
+MAX_VOLUME = 1.0
+MIN_VOLUME = 0.1
+CONTROLS_TIMEOUT = 5 # 5 seconds
+
+AMBIENT_DB = 30 # DB where MIN_VOLUME is applied
+DB_SCALE = 30 # AMBIENT_DB + DB_SCALE is where MAX_VOLUME is applied
+
+AudibleAlert = car.CarControl.HUDControl.AudibleAlert
+
+
+sound_list: Dict[int, Tuple[str, Optional[int], float]] = {
+  # AudibleAlert, file name, play count (none for infinite)
+  AudibleAlert.engage: ("engage.wav", 1, MAX_VOLUME),
+  AudibleAlert.disengage: ("disengage.wav", 1, MAX_VOLUME),
+  AudibleAlert.refuse: ("refuse.wav", 1, MAX_VOLUME),
+
+  AudibleAlert.prompt: ("prompt.wav", 1, MAX_VOLUME),
+  AudibleAlert.promptRepeat: ("prompt.wav", None, MAX_VOLUME),
+  AudibleAlert.promptDistracted: ("prompt_distracted.wav", None, MAX_VOLUME),
+
+  AudibleAlert.warningSoft: ("warning_soft.wav", None, MAX_VOLUME),
+  AudibleAlert.warningImmediate: ("warning_immediate.wav", None, MAX_VOLUME),
+}
+
+def check_controls_state(sm):
+  controls_missing = time.monotonic() - sm.rcv_time['controlsState']
+
+  if controls_missing > CONTROLS_TIMEOUT:
+    if sm['controlsState'].enabled and (controls_missing - CONTROLS_TIMEOUT) < 10:
+      return False
+
+  return True
+
+
+class Soundd:
+  def __init__(self):
+    self.load_sounds()
+
+    self.current_alert = AudibleAlert.none
+    self.current_volume = MIN_VOLUME
+    self.current_sound_frame = 0
+
+  def load_sounds(self):
+    self.loaded_sounds: Dict[int, np.ndarray] = {}
+
+    # Load all sounds
+    for sound in sound_list:
+      filename, play_count, volume = sound_list[sound]
+
+      wavefile = wave.open(BASEDIR + "/selfdrive/assets/sounds/" + filename, 'r')
+
+      assert wavefile.getnchannels() == 1
+      assert wavefile.getsampwidth() == 2
+      assert wavefile.getframerate() == SAMPLE_RATE
+
+      length = wavefile.getnframes()
+      self.loaded_sounds[sound] = np.frombuffer(wavefile.readframes(length), dtype=np.int16).astype(np.float32) / 32767
```

Comment:
what's 32767?

Reply:
2^16/2 for converting the int16 to float. I put that in